### ISSUE TITLE
Custom loss function

### ISSUE THREAD

**Issue by ogencoglu** [perpetual-ml/perpetual#36]
Ability to define arbitrary loss function just like in `lightgbm` would be great. Any such plans?

**Comment by deadsoul44** [comment]
It is part of the plan, but this might take a while. If you need any specific loss function, let me know.

**Comment by ogencoglu** [comment]
The ability to define different custom loss functions for different subsets of the training data would be my use case. I can do this with `lightgbm` (a bit hack-ish but gets the job done). My loss function is not constant as it is part of an optimization framework. Obviously gradients and Hessians should be available but there are frameworks for approximating those.

**Comment by jmrichardson** [comment]
@deadsoul44 A custom loss function would be really nice to try different scenarios.  Hoping this makes it to the top of priority :)  In mean time, can you create directional accuracy something like:

```
def directional_loss(y_true, y_pred):
    return -(np.sign(y_true) == np.sign(y_pred)).mean()
```

**Comment by linjing-lab** [comment]
src/objective_functions/mod.rs contains 4 different objectives, which covers most model fit situations,
```rust
pub enum Objective {
    LogLoss,
    SquaredLoss,
    QuantileLoss,
    AdaptiveHuberLoss
}
```

**Comment by deadsoul44** [comment]
Hello,

LogLoss is best for classification problems. 
SquaredLoss is best for regression problems. 
QuantileLoss is best for prediction intervals.
AdaptiveHuberLoss is best for handling outliers in regression. It is not released yet.

**Comment by deadsoul44** [comment]
v0.10.0 is released with custom objective support.

**Comment by ogencoglu** [comment]
Thank you! Is there a documentation with an example?
